contributors/design-proposals/core-metrics-pipeline.md

```diff
@@ -0,0 +1,315 @@
+
+The code that provides this data currently resides in cAdvisor.  I propose moving this to the kubelet as well.
+
+```go
+type CoreInfo struct {  
+  // MachineID reported by the node. For unique machine identification  
+  // in the cluster this field is prefered. Learn more from man(5)  
+  // machine-id: http://man7.org/linux/man-pages/man5/machine-id.5.html  
+  MachineID string `json:"machineID" protobuf:"bytes,1,opt,name=machineID"`  
+  // SystemUUID reported by the node. For unique machine identification  
+  // MachineID is prefered. This field is specific to Red Hat hosts  
+  // https://access.redhat.com/documentation/en-US/Red_Hat_Subscription_Management/1/html/RHSM/getting-system-uuid.html  
+  SystemUUID string `json:"systemUUID" protobuf:"bytes,2,opt,name=systemUUID"`  
+  // Boot ID reported by the node.  
+  BootID string `json:"bootID" protobuf:"bytes,3,opt,name=bootID"`  
+  // Kernel Version reported by the node from 'uname -r' (e.g. 3.16.0-0.bpo.4-amd64).  
+  KernelVersion string `json:"kernelVersion" protobuf:"bytes,4,opt,name=kernelVersion"`  
+  // OS Image reported by the node from /etc/os-release (e.g. Debian GNU/Linux 7 (wheezy)).  
+  OSImage string `json:"osImage" protobuf:"bytes,5,opt,name=osImage"`  
+  // Capacity represents the total resources of a node.  
+  // More info: http://kubernetes.io/docs/user-guide/persistent-volumes#capacity for more details.  
+  // +optional  
+  // ContainerRuntime Version reported by the node through runtime remote API (e.g. docker://1.5.0).  
+  ContainerRuntimeVersion string `json:"containerRuntimeVersion" protobuf:"bytes,6,opt,name=containerRuntimeVersion"`  
+  // Cloud provider the machine belongs to.  
+  CloudProvider CloudProvider `json:"cloud_provider"`  
+  // ID of cloud instance (e.g. instance-1) given to it by the cloud provider.  
+  InstanceID InstanceID `json:"instance_id"`  
+}  
+
+type CloudProvider string  
+
+const (  
+  GCE             CloudProvider = "GCE"  
+  AWS                           = "AWS"  
+  Azure                         = "Azure"  
+  Baremetal                     = "Baremetal"  
+  UnknownProvider               = "Unknown"  
+)
+```
+
+## Implementation Plan
+
+I will move all code pertaining to collection and processing of core metrics from cAdvisor into kubernetes.  
+I will vendor the new core metrics code back into cAdvisor. 
+I will modify volume stats collection so that it relies on this code.  
+I will modify the structure of stats collection code to be "On-Demand"   
+
+Tenative future work, not included in this proposal:  
+Obtain all runtime-specific information needed to collect metrics from the CRI.  
+Create a third party metadata API, whose function is to provide third party monitoring solutions with kubernetes-specific data (pod-container relationships, for example).  
+Modify cAdvisor to be "stand alone", and run in a seperate binary from the kubelet.  It will consume the above metadata API, and provide the summary API.  
```

Comment:
Suggested, i think.  I view this mostly as a worst-case proposal.  It may be that there is a better, and easier way to continue to provide the summary api, but at least providing stand-alone cAdvisor is a viable path that achieves the high level requirements.